packages/@blockprotocol/graph/src/type-resolver/entity-type-to-typescript.ts

```diff
@@ -0,0 +1,216 @@
+          read({ url }) {
+            const hardcodedType =
+              hardcodedBpTypes[url as keyof typeof hardcodedBpTypes];
+            if (hardcodedType) {
+              return hardcodedType;
+            }
+            return fetchTypeAsJson(url);
+          },
+        },
+      },
+    },
+    ...options,
+  });
+
+// A named type and the TS string defining it and its type dependencies
+type CompiledType = {
+  typeName: string;
+  typeScriptString: string;
+};
+
+// A map of schema URIs to their TS type name and definition
+type UriToType = { [typeUri: string]: CompiledType };
+
+/**
+ * Generates TypeScript types from a given Entity Type schema
+ * If depth > 0, follows 'links' from the schema to include types for link entities and their possible destinations
+ * @param schema the schema to generate types for
+ * @param depth the depth to which the graph of _entity type_ schemas linked from this schema will be resolved
+ * @param resolvedUrisToType a map of already-resolved schemas to skip type generation for
+ */
+const _jsonSchemaToTypescript = async (
+  schema: EntityType,
+  depth: number,
+  resolvedUrisToType: UriToType,
+): Promise<CompiledType & {}> => {
+  if (resolvedUrisToType[schema.$id]) {
+    return resolvedUrisToType[schema.$id]!;
+  }
+
+  // if the cache is empty, this is the schema the external caller provided. we need to know to add boilerplate up top
+  const rootSchema = Object.keys(resolvedUrisToType).length === 0;
+
+  const typeName = generateTypeNameFromSchema(schema);
+
+  if (!typeName) {
+    throw new Error("Schema is missing a title");
+  }
+
+  // This first generated type is just going to cover the entity's properties – they have other fields typed elsewhere
+  const propertyTypeName = `${typeName}Properties`;
+
+  // here we import the types defined elsewhere which we rely on, e.g. Entity
+  let compiledSchema = rootSchema ? generateImportStatements() : "";
+
+  // Generate the type for FooProperties, representing the entity's 'own properties' (not links or linked entities)
+  compiledSchema += await compileSchema(
+    { ...schema, title: propertyTypeName },
+    {
+      additionalProperties: false, // @todo add additionalProperties: false to entity type JSON schemas
+      bannerComment: rootSchema ? bannerComment(schema.$id, depth) : "",
+    },
+  );
+
+  // This adds 'type Foo = Entity<FooProperties>;'
+  const entityTypeDefinition = generateEntityDefinition(
+    typeName,
+    propertyTypeName,
+  );
+  compiledSchema += entityTypeDefinition;
+
+  // using a type instead of an interface allows using FooProperties in Entity<FooProperties>
+  // matching group needed in case the compiler has made e.g. FooProperties1 & FooProperties2 to handle inheritance
+  // @todo we need to catch types generated internally by the library via inheritance (allOf), which we don't know about
+  //   either these need to be converted to types too, or we need a different approach to Entity<Properties>
```

Comment:
I think we definitely need to convert them to types when we go about actually supporting inheritance.

Reply:
Agreed - I've made a subtask of the inheritance feature to fix this ([here](https://app.asana.com/0/0/1203642314378395/f), internal)